### DISCUSSION TITLE
Discussion about `server.accept()`

### DISCUSSION THREAD

**Discussion by JAndrassy** [JAndrassy/Arduino-Networking-API#2] [General]
Method [server:accept()](https://www.arduino.cc/reference/en/libraries/ethernet/server.accept/) was introduced by Paul Stoffregen in 2018 version 2 of the Arduino Ethernet library. I think it is a very useful addition.

The esp8266 WiFi library bundled with the esp8266 Arduino core was created in 2014 and method server.available was not implemented  as in the Ethernet and WiFi library. Its implementation behaves exactly like Paul's server.accept(). The ESP32 Arduino WiFi library copied this version of available() but added method accept() as alias. Recently esp8266 WiFi library added `accept()` as alias for `available()` too.

Networking libraries by Arduino, with exception of the Ethernet library, don't have `server.accept()`. WiFi and Ethernet libraries made by others have `server.accept()`, because it is really useful. 

In WiFiNINA my PR with accept() was approved but it is still not merged. The PR in the nina-fw was merged many months ago. If WiFiNINA gets server.accept(), other Arduino net libraries should have it too. My PR are waiting in WiFi101 and in WiFiS3 libraries.

What is you opinion on `server.accept()`?

**Comment by alranel** [comment]
I believe that accept() is quite useful so I agree with making it standard.

What worries me a bit is the non-standard implementation of `available()`.

**Reply by JAndrassy** [reply]
With the esp8266 Arduino core I achieved  with maintainers that in esp8266 core version 3+, WiFiServer.available() is deprecated with a big warning and accept() exists. I wrote for them a new server class names ArduinoWiFiServer, which has proper available() and print-to-all-client functionality. Their EthernetServer is alias to ArduinoWiFiServer for use with Ethernet drivers for LwIP there.

In esp32 core at least marking available() as deprecated would be good. They have accept() (as alias for available()). But in the past I had the feeling that they don't care much about good compatibility with Arduino. 

What I see on forums, people rarely need server.available as it works.  Most of the time accept() is what they need. Sometimes the behavior of available() is even in the way. 
Example: https://github.com/arduino-libraries/WiFiNINA/issues/176#issuecomment-1778911884

**Reply by alranel** [reply]
That sounds great. Only one strong recommendation: ArduinoWiFiServer is not a good name because according to the [library specification](https://arduino.github.io/arduino-cli/0.35/library-specification/#libraryproperties-file-format), only official libraries maintained by Arduino should be named with the "Arduino" prefix. I believe a better name would be WiFiServer2 or, since it's not limited to WiFi, you might want to consider GenericServer or something like that.

**Comment by alranel** [comment]
Meanwhile, this is a PR to add `accept()` to the Arduino mbed core (which supports many boards such as Portenta H7, GIGA R1 WiFi; Nano RP2040 Connect etc.):

https://github.com/arduino/ArduinoCore-mbed/pull/750

**Reply by JAndrassy** [reply]
perfect. 

I think the right way out of wrong implementations of server.available() is to 'push' accept(). Because available() is easy to implement with accept(). I have a generic template class which implements proper available() using accept() and array of Clients (and it does print-to-all-clients too of course). It is similar to esp8266 [ArduinoWiFiServer](https://github.com/esp8266/Arduino/blob/1662248b394740d82824f504a021d4dad3d314b6/libraries/ESP8266WiFi/src/ArduinoWiFiServer.h#L31) and I will publish it soon.

There are 4 examples for WiFiServer/EthernetServer. Two of them are traditional WebServer and ChatServer. Two are 'new': AdvancedChatServer and my PagerServer.

I created PagerServer to demonstrate available() and print-to-all-clients so people understand it.
Paul Stoffregen created AdvancedChatServer to demo server.accept().
WebServer example uses available() but would work the same with accept().
ChatServer was written to  demonstrate available(), but that was lost in later modifications.

**Reply by JAndrassy** [reply]
https://github.com/JAndrassy/NetApiHelpers#servertemplate
